src/borg/compress.pyx

```diff
@@ -142,7 +147,38 @@ class LZ4(CompressorBase):
             osize = LZ4_compress_default(source, dest, isize, osize)
         if not osize:
             raise Exception('lz4 compress failed')
-        return super().compress(dest[:osize])
+        # only compress if the result actually is smaller
+        if osize < isize:
+            return self, dest[:osize]
+        else:
+            return NONE_COMPRESSOR, None
+
+    def decide(self, data):
+        return self._decide(data)[0]
+
+    def decide_compress(self, data):
+        """
+        Decides what to do with *data* and handle accordingly. Returns (compressor, compressed_data).
+
+        *compressed_data* is the result of *data* being processed by *compressor*.
+        """
+        compressor, compressed_data = self._decide(data)
+
+        if compressor is self:
+            return self, super().compress(compressed_data)
+
+        if compressed_data is None:
+            compressed_data = compressor.compress(data)
+
+        return compressor, compressed_data
```

Comment:
considering what `self._decide` does, this could be just a simple `if ... return ... else return ...`.

Reply:
I did opt for the robust implementation depending less on how `_decide()` is actually implemented, covering the known special cases (compressor benig self and no compressed data being returned).

I see why you are leaning to a micro-optimization here instead, but do you still regard that the better choice if moving it to the `BaseCompressor`class, where a derived class could have its own `_decide()` that, e.g. similar to the one in auto compressor, may have more possible outcomes than the two cases of `(self, self_compressed_data)` and `(NONE_COMPRESSOR, None)`?

Reply:
yeah, you're right. let's keep the safer variant.